toke.c/regcomp.c: Simplified handling of new SVs

richardleach opened the pull request:
The 3 commits in this PR reflect the fact that calling generic function/macros on new SVs is often inefficient. 
* In `S_scan_heredoc`, `SvGROW` is replaced by `sv_grow_fresh` and `sv_setpvn` by `sv_setpvn_fresh`, depending upon the if/else branch.
* Also in `S_scan_heredoc`, `SvPVCLEAR` is inlined and simplified.
* In `S_regclass`, another `SvPVCLEAR` is inlined and simplified.

richardleach committed:
S_scan_heredoc: fresh sv functions close to point of use

S_scan_heredoc does a SvGROW on a fresh PVIV. A sv_grow_fresh is
more efficient, plus it seems only really needed in the nearby "else"
branch. In the "if" branch, sv_setsv_fresh can be used directly.

richardleach commented:
Yeah, that was my first thought but I wasn't sure if the low number of compatible uses justified it. I'm happy to do it for the reasoning you outlined though.

`SvPVCLEAR_FRESH` would keep with the `_fresh` naming convention used by other function in _sv.c_ and it can have the same constraints - only to be used on a freshly minted SVf_PV.

Sound okay?

demerphq commented:
Yeah, not gunna bike-shed the name, I just thought it was ironic as I am pretty sure I created the macro originally because i needed it in the regex engine, so possibly *because* of the macros you were making not use it. :-)

richardleach committed:
sv.c - add SvPVCLEAR_FRESH macro for new PVs with buffer

SvPVCLEAR(tmpstr) is a macro that calls sv_setpv_bufsize(tmpstr,0,0).
If called on a newly minted PV/PVIV/PVNV/PVMG that has a fresh PV
buffer assigned, many of the checks in SvPVCLEAR are unnecessary:
    * SvTHINKFIRST - won't be set
    * upgrade to SVt_PV - won't be necessary
    * SvGROW - we assert that this isn't necessary

It then does some useful, but small actions:
    * sets SvCUR(sv, 0);
    * sets the last byte in the PV buffer to \0
    * sets SvPOK_only_UTF8
    * does SvTAINT

Then does a final irrelevant check:
    * is tmpstr a PVCV - no it is not

This commit introduces SvPVCLEAR_FRESH as a drop-in replacement
for compatible SVs, that only does the relevant small actions.

richardleach committed:
S_scan_heredoc - use SvPVCLEAR_FRESH on new SV

richardleach committed:
S_regclass - use SvPVCLEAR_FRESH on new SV

richardleach committed:
vnewSVpvf - inline & reduce call to sv_vsetpvfn

sv_vsetpvfn does just two things:
* SvPVCLEAR on the sv passed in
* calls sv_vcatpvfn_flags

Since the SV passed into sv_vsetpvfn by vnewSVpvf is a brand new SV,
it is more efficient to inline sv_vsetpvfn and change the SvPVCLEAR
to SvPVCLEAR_FRESH.

richardleach commented:
Ok, I've revised this PR with a new SvPVCLEAR_FRESH.

Also moved #20274 into this PR as an extra commit, since the new macro makes that change more straightforward.

richardleach commented:
@demerphq - you okay with this PR now?

demerphq commented:
Looka good to me.

Thanks!

Yves

On Sat, 17 Sept 2022, 23:16 Richard Leach, ***@***.***> wrote:

> @demerphq <https://github.com/demerphq> - you okay with this PR now?
>
> —
> Reply to this email directly, view it on
> <https://github.com/Perl/perl5/pull/20275#issuecomment-1250148531>
> Message ID: ***@***.***>
>
